### PR TITLE
perf(ngAnimate): avoid repeated calls to addClass/removeClass when an…

### LINKED ISSUE #14165: ngAnimate doesn't cache failed animations to avoid addClass/removeClass
**Note: for support questions, please use one of these channels:** https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#question. This repository's issues are reserved for feature requests and bug reports.
- **Do you want to request a _feature_ or report a _bug_?**
- **What is the current behavior?**
- **If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem** via
  https://plnkr.co or similar (you can use this template as a starting point: http://plnkr.co/edit/tpl:yBpEi4).
- **What is the expected behavior?**
- **What is the motivation / use case for changing the behavior?**
- **Which version of Angular, and which browser and OS does this issue affect?** Did this work in previous
  versions of Angular? Please also test with the latest stable and snapshot versions.
- **Other information** (e.g. detailed explanation, stacktraces, related issues, suggestions how to fix)

### PR DESCRIPTION
**Pull request by Narretz**
…imation has no duration

Background:
ngAnimate writes helper classes to DOM elements to see if animations are defined on them. If many
elements have the same definition, we can cache the definition and skip the application of the
helper classes altogether. This helps particularly with large ngRepeat collections.

Closes #14165
Closes #14166

**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**

**What is the current behavior? (You can also link to an open issue here)**

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change?**

**Please check if the PR fulfills these requirements**
- [ ] The commit message follows our [guidelines](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#commits)
- [ ] Fix/Feature: [Docs](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#documentation) have been added/updated
- [ ] Fix/Feature: Tests have been added; existing tests pass

**Other information**:

### CODE REVIEW

File: src/ngAnimate/animateCache.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+var $$AnimateCacheProvider = function() {
+
+  var KEY = '$$ngAnimateParentKey';
+  var parentCounter = 0;
+  var cache = Object.create(null);
+
+  this.$get = [function() {
+    return {
+      cacheKey: function(node, method, addClass, removeClass) {
+        var parentNode = node.parentNode;
```
Comment (gkalpak) [inline src/ngAnimate/animateCache.js]: Isn't it possible for `node.parentNode` to be undefined?

File: src/ngAnimate/shared.js [hunk 2]
```diff
@@ -314,7 +314,7 @@ function applyGeneratedPreparationClasses(element, event, options) {
   }
   if (classes.length) {
     options.preparationClasses = classes;
-    element.addClass(classes);
+    $$jqLite.addClass(element, classes);
```
Comment (gkalpak) [inline src/ngAnimate/shared.js]: Why this change?

File: src/ngAnimate/animateCache.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        var entry = cache[key];
+        return entry ? entry.total : 0;
+      },
+
+      get: function(key) {
+        var entry = cache[key];
+        return entry && entry.value;
+      },
+
+      put: function(key, value, isValid) {
```
Comment (gkalpak) [inline src/ngAnimate/animateCache.js]: What happens when you can `put` multiple times for the same key with different `isValid` values?
Why isn't `isValid` updated?

File: src/ngAnimate/animateCss.js [hunk 4]
```diff
@@ -377,20 +340,26 @@ var $AnimateCssProvider = ['$animateProvider', /** @this */ function($animatePro
         }
       }
 
+      // if a css animation has no duration we
+      // should mark that so that repeated addClass/removeClass calls are skipped
+      var hasNoDuration = allowNoDuration || (timings.transitionDuration > 0 || timings.animationDuration > 0);
```
Comment (gkalpak) [inline src/ngAnimate/animateCss.js]: Shouldn't this be named `hasDuration`?

File: src/ngAnimate/animateCss.js [hunk 5]
```diff
@@ -903,10 +879,10 @@ var $AnimateCssProvider = ['$animateProvider', /** @this */ function($animatePro
           $$jqLite.addClass(element, activeClasses);
 
           if (flags.recalculateTimingStyles) {
-            fullClassName = node.getAttribute('class') + ' ' + preparationClasses;
-            cacheKey = gcsHashFn(node, fullClassName);
+            fullClassName = node.className + ' ' + preparationClasses;
```
Comment (gkalpak) [inline src/ngAnimate/animateCss.js]: [According to MDN][1]:

> `className` can also be an instance of `SVGElement` if the element is an `SVGAnimatedString`. It is better to get/set the `className` of an element using `Element.getAttribute` and `Element.setAttribute` if you are dealing with SVG elements.

[1]: https://developer.mozilla.org/en-US/docs/Web/API/Element/className

File: src/ngAnimate/animation.js [hunk 6]
```diff
@@ -221,7 +237,32 @@ var $$AnimationProvider = ['$animateProvider', /** @this */ function($animatePro
...
+          var innerArray = finalAnimations[i];
+          for (var j = 0; j < innerArray.length; j++) {
+            var entry = innerArray[j];
+            var element = entry.element;
+
+            // the RAFScheduler code only uses functions
+            finalAnimations[i][j] = entry.fn;
+
+            // the first row of elements shouldn't have a prepare-class added to them
+            // since the elements are at the top of the animation hierarcy and they
```
Comment (gkalpak) [inline src/ngAnimate/animation.js]: _hierarcy_ --> hierarchy_

File: src/ngAnimate/animateCache.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        var parentNode = node.parentNode;
+        var parentID = parentNode[KEY] || (parentNode[KEY] = ++parentCounter);
+        var parts = [parentID, method, node.getAttribute('class')];
+        if (addClass) {
+          parts.push(addClass);
+        }
+        if (removeClass) {
+          parts.push(removeClass);
+        }
+        return parts.join(' ');
```
Comment (gkalpak) [inline src/ngAnimate/animateCache.js]: Is it a concerns that adding/removing the same class can result in the same key?
E.g. `cacheKey(node, merhod, 'foo') === cacheKey(node, method, null, 'foo')`.

**Review by gkalpak — COMMENTED** [review]
I won't pretend to understand everything that's going on, but looks reasonable (and there are tests that test stuff :grin:)

**Review by Narretz — COMMENTED** [review]
Thanks for the thorough review, @gkalpak I don't think the points are actionable, but it's possible that I missed something.

File: test/ngAnimate/animateCacheSpec.js [hunk 8]
```diff
@@ -73,24 +73,27 @@ describe('ngAnimate $$animateCache', function() {
     expect($$animateCache.count('key2')).toBe(0);
   }));
 
-  it('should store whether the stored entry is invalid or not', inject(function($$animateCache) {
-    var validEntry = { someEssentialProperty: true };
-    var invalidEntry = { someEssentialProperty: false };
+  describe('containsCachedAnimationWithoutDuration', function() {
+    it('should return false if the validity of an key is false', inject(function($$animateCache) {
```
Comment (gkalpak) [inline test/ngAnimate/animateCacheSpec.js]: _an key_ --> _a key_

**Review by gkalpak — APPROVED** [review]
Other than the typos, LGTM 👍
